src/mlpack/core/dists/gamma_distribution.cpp

```diff
@@ -0,0 +1,76 @@
+/**
+ * @file gamma_distribution.cpp
+ * @author Yannis Mentekidis
+ *
+ * Implementation of the methods of GammaDistribution.
+ */
+#include "gamma_distribution.hpp"
+
+#include <boost/math/special_functions/digamma.hpp>
+//#include <boost/math/special_functions/trigamma.hpp> // Moved to prereqs.hpp
+
+using namespace mlpack;
+using namespace mlpack::distribution;
+
+
+// Returns true if computation converged.
+inline bool GammaDistribution::converged(double aOld, double aNew)
+{
+  return (std::abs(aNew - aOld) / aNew) < tol;
+}
+
+// Fits an alpha and beta parameter to each dimension of the data.
+void GammaDistribution::Train(const arma::mat& rdata)
+{
+  // Use pseudonym fittingSet regardless of if rdata was provided by user.
+  const arma::mat& fittingSet = 
+    rdata.n_elem == 0 ? referenceSet : rdata;
+
+  // If fittingSet is empty, nothing to do.
+  if (arma::size(fittingSet) == arma::size(arma::mat()))
+    return;
+
+  // Use boost's definitions of digamma and tgamma, and std::log.
+  using boost::math::digamma;
+  using boost::math::trigamma;
+  using std::log;
+
+  // Allocate space for alphas and betas (Assume independent rows).
+  alpha.set_size(fittingSet.n_rows);
+  beta.set_size(fittingSet.n_rows);
+
+  // Treat each dimension (i.e. row) independently.
+  for (size_t row = 0; row < fittingSet.n_rows; ++row)
+  {
+    // Calculate log(mean(x)) and mean(log(x)) required for fitting process.
+    const double meanLogx = arma::mean(arma::log(fittingSet.row(row)));
+    const double meanx = arma::mean(fittingSet.row(row));
+    const double logMeanx = std::log(arma::mean(meanx));
```

Comment:
Hm, if you are willing to pay a little more memory to calculate the means of each dimension all at once, instead of one row at a time, I think that you can get some speedup here, since row-wise access is suboptimal (but accessing the full matrix to calculate the mean should be fine).  Then inside this loop where you estimate parameters for each dimension, you could just use the relevant mean/log-mean/mean-log from the vector that was calculated.